### PR TITLE
Prototype solidity implementation

### PR DESCRIPTION
**Pull request by lalexgap**
Attempt to implement a minimal prototype that serve as proof of concept and a reference the submission statement.

### CODE REVIEW

File: contracts/L2.sol [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        );
+
+        entry.receiver.transfer(entry.value);
+    }
+
+    function lockFundsInEscrow(
+        address payable receiver,
+        bytes32 escrowHash,
+        uint256 payoutDate,
+        uint256 reclaimDate
```
Comment (geoknee) [inline contracts/L2.sol:66]: I feel there should probably be some input validation on `payoutDate` and `reclaimDate`? Should one proceed the other? And by some minimum  or maximum amount? 

I'd also consider using `block.number` instead of `block.timestamp` https://medium.com/@phillipgoldberg/smart-contract-best-practices-revisited-block-number-vs-timestamp-648905104323

File: contracts/L1.sol [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            ticketSigner == ticket.sender,
+            "Ticket is not signed by sender"
+        );
+        require(
+            balances[ticket.sender] >= ticket.value,
+            "Sender does not have enough funds"
+        );
+
+        currentNonce++;
+        ticket.receiver.transfer(ticket.value);
```
Comment (geoknee) [inline contracts/L1.sol:62]: Don't you need to decrease the balance of the sender? For example:
```suggestion
        balances[ticket.sender] -= ticket.value
        ticket.receiver.transfer(ticket.value);
```

File: contracts/L2.sol [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            commitedTicket.nonce == secondTicket.nonce,
+            "The two tickets must have the same nonce."
+        );
+
+        EscrowEntry memory entry = funds[msg.sender];
+
+        require(
+            entry.escrowHash == keccak256(abi.encode(escrowSecret)),
+            "Invalid preimage."
+        );
```
Comment (geoknee) [inline contracts/L2.sol:140]: I think this method will succeed (not revert) if I just set `secondTicket = commitedTicket` and `secondSignature = firstSignature`. Is that desired behaviour?

File: contracts/L2.sol [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            "Cannot add an entry for a different address"
+        );
+        funds[receiver] = EscrowEntry(
+            receiver,
+            payable(msg.sender),
+            msg.value,
+            payoutDate,
+            reclaimDate,
+            escrowHash
+        );
```
Comment (geoknee) [inline contracts/L2.sol:81]: I'm not sure the check on L70 is right. If this is the first time someone is locking funds, `entry.receiver = funds[receiver].receiver = 0` so this check will never pass.

I think you're covered by the use of `msg.sender` on L76, so you can just remove the check above?

In other words, you need to enforce the correct *write* of this field, and do not need to do a *read* of it beforehand.

File: contracts/L2.sol [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    // This is a timestamp of when the escrow can be claimed by the receiver
+    uint256 payoutDate;
+    // This is a timestamp of when the funds can reclaimed back to the original sender
+    uint256 reclaimDate;
+    // This is the hash of some secret preimage.
+    bytes32 escrowHash;
+}
+
+contract L2Contract {
+    mapping(address => EscrowEntry) private funds;
```
Comment (geoknee) [inline contracts/L2.sol:28]: Is this a better name?
```suggestion
    mapping(address => EscrowEntry) private escrowEntries;
```
The fewer distinct names in play, the better (IMO).

File: contracts/L1.sol [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+contract L1Contract {
+    uint256 currentNonce = 0;
+    mapping(address => uint256) balances;
+
+    function claimTickets(
+        WithdrawalTicket[] calldata tickets,
+        bytes32[] calldata escrowPreimages,
+        bytes32[] calldata r,
+        bytes32[] calldata s,
+        uint8[] calldata v
```
Comment (geoknee) [inline contracts/L1.sol:21]: This would be a bit nicer as `Signature[] sigs`, where `Signature` is a struct containing a `v,r,s`. See #2 (and `nitro-protocol`).

File: contracts/L2.sol (hunk 3 above)
Comment (geoknee) [reply contracts/L2.sol:140]: See #2

File: contracts/L2.sol (hunk 1 above)
Comment (lalexgap) [reply contracts/L2.sol:66]: Right now there's an implicit assumption that Alice will not provide the preimage if Bob creates a Ticket with unreasonable timeouts. I don't see the harm in hardcoding some sanity checks though.

I also think using `block.number` might not be the right fit in this case? The medium article warns that "extrapolating block.number to a time is dangerous". It seems if we use `block.number` we'd be using a timeout specified by the amount of blocks, which we'd extrapolate into a number when talking about the protocol.

File: contracts/L2.sol (hunk 3 above)
Comment (lalexgap) [reply contracts/L2.sol:140]: Ah good catch!

File: contracts/L1.sol (hunk 6 above)
Comment (lalexgap) [reply contracts/L1.sol:21]: Yup, that makes sense. I went with the simplest solution since I wasn't sure how hard encoding the signatures as structs would be for the client side code.

File: contracts/L2.sol (hunk 3 above)
Comment (lalexgap) [reply contracts/L2.sol:150]: Fixed in 1c40a31

File: contracts/L1.sol (hunk 6 above)
Comment (lalexgap) [reply contracts/L1.sol]: Done in 2e18bbf

File: contracts/L2.sol (hunk 5 above)
Comment (lalexgap) [reply contracts/L2.sol]: Done in a3fc71f605c4c25f72239352217512ed1d8203c2

File: contracts/L2.sol [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            receiver,
+            payable(msg.sender),
+            msg.value,
+            escrowExpiry,
+            reclaimDate,
+            escrowHash
+        );
+    }
+
+    mapping(address => uint256) nonces;
```
Comment (kerzhner) [inline contracts/L2.sol:94]: What is the reason we need to keep track of nonces on L2?

File: contracts/L2.sol [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        require(
+            ticket.nonce == nonces[ticket.sender] + 1,
+            "The ticket must use the next nonce."
+        );
+
+        bytes32 ticketHash = keccak256(abi.encode(ticket));
+        address ticketSigner = recoverSigner(ticketHash, ticketSignature);
+
+        require(
+            ticket.sender == ticketSigner,
```
Comment (kerzhner) [inline contracts/L2.sol:111]: What do you think about removing this restriction? It seems that the ticket signer can be different than the ticket sender as long as all book keeping is done with ticket signer as mapping keys.

File: contracts/L2.sol [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+    function lockFundsInEscrow(
+        // Called by Alice, with receiver = Bob.
+        // It's a hash-time-lock, with a window of opportunity for Bob.
+        // Alice will pull her funds out unless she gets funds from Bob on L1 before reclaimDate
+        address payable receiver,
+        bytes32 escrowHash,
+        uint256 escrowExpiry,
+        uint256 reclaimDate
+    ) public payable {
```
Comment (kerzhner) [inline contracts/L2.sol:77]: Are funds actually getting locked up in this method?

File: contracts/L2.sol [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    bytes32 r;
+    bytes32 s;
+    uint8 v;
+}
+
+struct WithdrawalTicket {
+    uint256 value;
+    uint256 nonce;
+    address payable receiver;
+    address sender;
```
Comment (kerzhner) [inline contracts/L2.sol:14]: Do you think this field is needed?

File: contracts/L2.sol [hunk 11]
```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+struct Signature {
```
Comment (kerzhner) [inline contracts/L2.sol:4]: Should `Signature` and `WithdrawalTicket` get factored out into a base contract?

File: contracts/L1.sol [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    uint256 value;
+    uint256 nonce;
+    address payable receiver;
+    address sender;
+    bytes32 escrowHash;
+}
+
+contract L1Contract {
+    mapping(address => uint256) nonces;
+    mapping(address => uint256) balances;
```
[the remark this answers is not in the document]
Comment (lalexgap) [reply contracts/L1.sol:19]: Ah shoot! I think George caught this and I had applied his suggestion but then lost it in a rebase.

TLDR: This if fixed in eb60cf384721fa6fd71ec2ec4ed7c2e80d2bc144

File: contracts/L2.sol (hunk 9 above)
Comment (lalexgap) [reply contracts/L2.sol:77]: https://github.com/statechannels/fast-exit/pull/1#discussion_r750785646

File: contracts/L2.sol [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    function lockFundsInEscrow(
+        // Called by Alice, with receiver = Bob.
+        // It's a hash-time-lock, with a window of opportunity for Bob.
+        // Alice will pull her funds out unless she gets funds from Bob on L1 before reclaimDate
+        address payable receiver,
+        bytes32 escrowHash,
+        uint256 escrowExpiry,
+        uint256 reclaimDate
+    ) public payable {
+        EscrowEntry memory entry = escrowEntries[receiver];
```
[the remark this answers is not in the document]
Comment (lalexgap) [reply contracts/L2.sol:78]: This was mostly done to simplify my mental model :P

 I think there could multiple transfers with the `receiver`. Bob just has to keep track of all tickets for Alice to know how much of Alice's escrow funds are accounted for before he authorizes a new ticket.

File: contracts/L2.sol (hunk 11 above)
Comment (lalexgap) [reply contracts/L2.sol:4]: Yup!

File: contracts/L2.sol (hunk 8 above)
Comment (lalexgap) [reply contracts/L2.sol:111]: Would there be a mapping from ticket signer to ticket sender stored on chain? Otherwise couldn't anyone submit a ticket committing someone else to sending funds?

Comment (kerzhner) [reply contracts/L2.sol:111]: Yep, we would store the a mapping of signer to all of the `WithdrawalTickets` the signer has authorized.

File: contracts/L2.sol [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        bytes32 ticketHash = keccak256(abi.encode(ticket));
+        address ticketSigner = recoverSigner(ticketHash, ticketSignature);
+
+        require(
+            ticket.sender == ticketSigner,
+            "The ticket must be signed by the sender."
+        );
+
+        nonces[ticket.sender]++;
+        ticketCommitments[ticket.sender] = ticketHash;
```
Comment (kerzhner) [inline contracts/L2.sol:116]: Is it true that a signer (or sender) is likely to authorize many tickets at a time?

File: contracts/L2.sol [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+    mapping(address => uint256) nonces;
+    mapping(address => bytes32) ticketCommitments;
+
+    function commitToWithdrawal(
+        // Called by Bob, with the ticket.receiver = Alice. He is commiting to an L1 withdrawal
+        WithdrawalTicket calldata ticket,
+        Signature calldata ticketSignature
+    ) public {
+        require(
+            ticket.nonce == nonces[ticket.sender] + 1,
```
Comment (kerzhner) [inline contracts/L2.sol:103]: Why do we need to keep track of nonces in this contract? Does this provide any safety?

File: contracts/L2.sol (hunk 9 above)
Comment (kerzhner) [reply contracts/L2.sol:77]: That comment is around balances on L1. Does it also apply to escrowed funds on L2?

**Review by kerzhner — APPROVED** [review]
Looks good to me in general. Nice work translating our design doc to the prototype!

File: contracts/L2.sol (hunk 8 above)
Comment (kerzhner) [reply contracts/L2.sol:111]: My general mental model is that `WithdrawalTickets` are per signer. And later, a signer will be penalized for signing "duplicate nonce" tickets.

Comment (lalexgap) [reply contracts/L2.sol:111]: I think that approach makes sense and makes the protocol more flexible! I'd be in favour of doing that in a separate PR

File: contracts/L2.sol (hunk 15 above)
Comment (lalexgap) [reply contracts/L2.sol:103]: I think the L2 contract needs to check the ticket Bob is committing to has a nonce one higher than the previous ticket he committed to. 

Right now since we store the complete `EscrowEntry` in `storage` I don't we think we need a seperate `nonce` entry as we can just look up the previous `EscrowEntry` and check that the the new ticket's nonce is larger.

However if we start storing the `EscrowEntry` hash (which we should!) I think we'll need to store `nonces` separately so we can do this check?

File: contracts/L2.sol (hunk 14 above)
Comment (lalexgap) [reply contracts/L2.sol:116]: Yup, that's a good point!

We should be storing ticket commitments per sender **and** nonce. That way a sender can easily authorize many tickets at once.

File: contracts/L2.sol (hunk 11 above)
Comment (lalexgap) [reply contracts/L2.sol]: Done in 20b96be
